### PR TITLE
feat(Forms): add `country` prop to Field.PostalCodeAndCity to support other countries than Norway

### PR DESCRIPTION
**Pull request by joakbjerk**
- [x] Add example
- [x] Document props
- [x] Add tests
- [x] Add path support for `country` prop
- [x] Figure out wonky behavior in new tests 🤔
- [x] Fix `mask` type error
- [x] Add demo for path based country values
- [x] Update Postal Code input width in example

### CODE REVIEW

File: packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/demos.mdx [hunk 1]
```diff
@@ -44,6 +44,12 @@ By using the `itemPath` property, you can iterate over an array and use the `pos
 
 <Examples.ValidationRequired />
 
+### Non Norwegain Postal Codes
```
Comment (boyum) [inline packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/demos.mdx:47]:
```suggestion
### Non-Norwegian Postal Codes
```

File: packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/demos.mdx [hunk 2]
```diff
@@ -44,6 +44,12 @@ By using the `itemPath` property, you can iterate over an array and use the `pos
...
+
+If you want to allow for a postal code that is not Norwegain, just set the `country` property to the desired country and add your own custom validation.
```
Comment (boyum) [inline packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/demos.mdx:49]:
```suggestion
If you want to allow for a postal code that is not Norwegian, just set the `country` property to the desired country and add your own custom validation.
```

**Comment by tujoworker** [comment]
> How to fix? refresh cache? Close the CLI and browser and restart these app and processes did work for me when I saw such an error.

File: packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx [hunk 3]
```diff
@@ -1,26 +1,67 @@
...
 function PostalCodeAndCity(props: Props) {
   const translations = useTranslation()
+  const { getSourceValue } = useDataValue()
 
   const {
     postalCode = {},
     city = {},
     help,
     width = 'large',
+    country = 'no',
```
[the remark this answers is not in the document]
Comment (tujoworker) [reply packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx:32]: We have a file called `defaults.ts`. But we don't have a country default there. Maybe we can/should add:

```ts
export const COUNTRY = 'NO' // ISO 3166-1 alpha-2
```

File: packages/dnb-eufemia/src/extensions/forms/Value/PostalCodeAndCity/PostalCodeAndCity.tsx [hunk 4]
```diff
@@ -40,5 +33,17 @@ function PostalCodeAndCity(props: Props) {
...
+  let value = props.value ?? undefined
+
+  const valueProps = useValueProps(props)
+
+  if (valueProps.value) {
+    value = valueProps.value
+  }
+
+  return value
+}
```
Comment (tujoworker) [inline packages/dnb-eufemia/src/extensions/forms/Value/PostalCodeAndCity/PostalCodeAndCity.tsx:46]:
```suggestion
function usePostalCodeAndCityValue(props: Props) {
  return useValueProps(props).value ?? props.value
}
```

This should be equivalent with less code.

**Comment by langz** [comment]
Looks good, I'll do a quick round of testing after lunch 🐔

File: packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/demos.mdx [hunk 5]
```diff
@@ -44,6 +44,18 @@ By using the `itemPath` property, you can iterate over an array and use the `pos
 
 <Examples.ValidationRequired />
 
+### Path Based Country
+
+The `country` prop supports a field path as value. This allows you to set the `country` based on the value of another field.
```
Comment (langz) [inline packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/demos.mdx:49]:
```suggestion
The `country` property supports a field path as value. This allows you to set the `country` based on the value of another field.
```

File: packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx [hunk 6]
```diff
@@ -1,26 +1,71 @@
...
+  )
+
+  const postalCodeValidationProps = useMemo(() => {
+    return {
+      mask: postalCode?.mask
+        ? postalCode.mask
+        : isNorway
+        ? [/\d/, /\d/, /\d/, /\d/]
+        : undefined,
+      pattern: postalCode.pattern
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx:51]:
```suggestion
      pattern: postalCode?.pattern
```

Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx:51]:
```suggestion
      pattern: postalCode?.pattern
```

**Comment by langz** [comment]
I added a few minor commits, which is just "code styling", and could be discussed if it improves the code-readability or not 😂 

Anyhow, I like this, and it works good :) 

# I have a few thoughts

## In the given example(Non-Norwegian Postal Codes) why would I as a consumer want to set the country?
Screenshot 2024-10-16 at 12 38 21

## The spacing between the input fields in the component(PostalCodeAndCity) when displaying country is a bit tight/crammed vertically. Should it be so? It perhaps makes sense that it's a bit tighter vertically spacing here, compared to if it would just be three separate fields(not put together in a component with a given purpose):
Screenshot 2024-10-16 at 12 39 14

**Comment by joakbjerk** [comment]
> I added a few minor commits, which is just "code styling", and could be discussed if it improves the code-readability or not 😂
> 
> Anyhow, I like this, and it works good :)
> ### I have a few thoughts
> #### In the given example(Non-Norwegian Postal Codes) why would I as a consumer want to set the country to be "de"?
> Screenshot 2024-10-16 at 12 38 21
> #### The spacing between the input fields in the component(PostalCodeAndCity) when displaying country is a bit tight/crammed vertically. Should it be so? It perhaps makes sense that it's a bit tighter vertically spacing here, compared to if it would just be three separate fields(not put together in a component with a given purpose), but it still feels a bit tight:
> Screenshot 2024-10-16 at 12 39 14

Added space between the fields in demo 👍

The `country` property thingy stems from [this slack thread](https://dnb-it.slack.com/archives/C04NL73S2Q5/p1728472603329999), where it was discussed that setting the country prop to something other than `NO` would remove the default Norwegian validation used by the component. But as you say, it's not clear why one would to that, as the property really just functions as a kinda of unclear `removeDefaultValidation` property, since we do not provide any built in validation rules based on the value of the country prop itself 🤷‍♂️

**Comment by tujoworker** [comment]
> The spacing between the input fields in the component(PostalCodeAndCity) when displaying country is a bit tight/crammed vertically. Should it be so?

Maybe there's a more declarative way of doing this. But I think, but having the country prop, it allows us eventually to progressively add support for other countries. But it could be done via a "plugin" system that a dev can import and provide for certain countries, just to lower the amount of extra code it is needed to support many other countries. 

But the most important thing to support is removing the limitation – when "myCountryPath" has changed.

File: packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx [hunk 7]
```diff
@@ -41,30 +41,31 @@ function PostalCodeAndCity(props: Props) {
...
+  const {
+    pattern: cityPattern,
+    className: cityClassName,
+    label: cityLabel,
+    width: cityWidth,
+    errorMessages: cityErrorMessages,
+  } = city
+
+  const {
+    mask: postalCodeMask = isNorway ? [/\d/, /\d/, /\d/, /\d/] : undefined,
```
Comment (langz) [inline packages/dnb-eufemia/src/extensions/forms/Field/PostalCodeAndCity/PostalCodeAndCity.tsx]: Comment for reference

**Comment by tujoworker** [comment]
For some reason the placeholder does not update:
Screenshot 2024-10-16 at 13 25 16

Also, the size is too large. We should consider to change the default size. Just a view pixels are missing:

Screenshot 2024-10-16 at 13 23 16

File: packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/Examples.tsx [hunk 8]
```diff
@@ -172,3 +172,35 @@ export const IterateArray = () => {
...
 }
+
+export const SettingCountryBasedOnPath = () => {
+  return (
+    <ComponentBox>
+      <Form.Handler>
+        <Field.SelectCountry
+          path="/myCountry"
+          defaultValue="NO"
+          bottom="medium"
```
Comment (tujoworker) [inline packages/dnb-design-system-portal/src/docs/uilib/extensions/forms/feature-fields/PostalCodeAndCity/Examples.tsx:183]:
```suggestion
```

We should use Flex.Stack or a card with stack: `<Card stack>`

**Comment by joakbjerk** [comment]
> For some reason the placeholder does not update: Screenshot 2024-10-16 at 13 25 16
> 
> Also, the size is too large. We should consider to change the default size. Just a view pixels are missing:
> Screenshot 2024-10-16 at 13 23 16

Should be working again now ⭐️

**Comment by joakbjerk** [comment]
> > For some reason the placeholder does not update: Screenshot 2024-10-16 at 13 25 16
> > Also, the size is too large. We should consider to change the default size. Just a view pixels are missing:
> > Screenshot 2024-10-16 at 13 23 16

Should be working again now ⭐️

**Comment by joakbjerk** [comment]
> For some reason the placeholder does not update: Screenshot 2024-10-16 at 13 25 16
> 
> Also, the size is too large. We should consider to change the default size. Just a few pixels are missing:
> Screenshot 2024-10-16 at 13 23 16

Regarding the width of the input, I agree, it's to wide, but do we have an option to make width on the input accommodate the mask, or placeholder length? I've tried setting the size prop to the mask length, like you could with the masked input, without any luck ☹️

**Comment by tujoworker** [comment]
Added this PR #4142 which we can use to set a larger width on the postal code field.

**Comment by joakbjerk** [comment]
With the new width property:

Screenshot 2024-10-18 at 10 21 54

**Review by langz — APPROVED** [review]
Looks great 🚀

**Review by tujoworker — APPROVED** [review]

**Comment by tujoworker** [comment]
Nice work 🥁

**Comment by tujoworker** [comment]
My last commit changes how we handle the defaults. I think, this way we are more flexible to adjust specific needs in the future. Also, for most other countries, the size will be larger of the postal code field.

Screenshot 2024-10-18 at 11 10 49

**Pull request merged**